operator/pkg/dynamicwatch/watch.go

```diff
@@ -0,0 +1,337 @@
+	}
+	return annotations
+}
+
+func getLabelsOrEmptyMap(meta metav1.Object) map[string]string {
+	labels := meta.GetLabels()
+	if labels == nil {
+		labels = map[string]string{}
+	}
+	return labels
+}
+
+//
+// Functions for loading and saving watchers in an annotation
+// 
+
+func (impl *DynamicWatchesAnnotationImpl) getWatcherNames(meta metav1.Object) []string {
+	annotations := getAnnotationsOrEmptyMap(meta)
+	content, ok := annotations[WatchedByAnnotation]
+
+	if !ok {
+		content = ""
+	}
+	
+	data := []string{}
+	if content != "" {
+		err := json.Unmarshal([]byte(content), &data)
+		if err != nil {
+			impl.Logger.Error(err, "Failed to parse watchers data", 
+				"watched", namespacedNameString(meta))
+
+			// As opposed to erroring out here, we'll just allow the
+			// annotation to be replaced with a valid one
+			data = []string{}
+		}
+	}
+
+	return data
+}
+
+func (impl *DynamicWatchesAnnotationImpl) updateWatcherNames(meta metav1.Object, names []string) bool {
+	originalWatchers := impl.getWatcherNames(meta)
+	originalHasWatchedLabel := hasWatchedLabel(meta)
+
+	if len(names) == 0 {
+		annotations := getAnnotationsOrEmptyMap(meta)
+		delete(annotations, WatchedByAnnotation)
+		meta.SetAnnotations(annotations)
+
+		labels := getLabelsOrEmptyMap(meta)
+		delete(labels, WatchedLabel)
+		meta.SetLabels(labels)
+	} else {
+		bytes, err := json.Marshal(names)
+
+		if err != nil {
+			impl.Logger.Error(err, "Failed to updated watchers on watched resource", "watched", namespacedNameString(meta))
+		} else {
+			annotations := getAnnotationsOrEmptyMap(meta)
+			annotations[WatchedByAnnotation] = string(bytes)
+			meta.SetAnnotations(annotations)
+			labels := getLabelsOrEmptyMap(meta)
+			labels[WatchedLabel] = "true"
+			meta.SetLabels(labels)
+		}
+	}
+	newWatchers := impl.getWatcherNames(meta)
+	newHasWatchedLabel := hasWatchedLabel(meta)
+	return !reflect.DeepEqual(originalWatchers, newWatchers) || originalHasWatchedLabel != newHasWatchedLabel
+}
+
+//
+// Functions for manipulating watchers in annotation
+//
+
+func (impl *DynamicWatchesAnnotationImpl) removeWatcher(meta metav1.Object, watcher string) bool {
+	watchers := impl.getWatcherNames(meta)
+	watchers = utils.RemoveValueFromStringArray(watchers, watcher)
+	changedWatcherValues := impl.updateWatcherNames(meta, watchers)
+
+	return changedWatcherValues
+}
+
+func (impl *DynamicWatchesAnnotationImpl) addWatcher(meta metav1.Object, watcher string) bool {
+	watchers := impl.getWatcherNames(meta)
+	watchers = utils.AppendValuesToStringArrayIfNotPresent(watchers, watcher)
+	return impl.updateWatcherNames(meta, watchers)
+}
+
+func hasWatchedLabel(meta metav1.Object) bool {
+	labels := getLabelsOrEmptyMap(meta)
+	value, ok := labels[WatchedLabel]
+	return ok && "true" == value
+}
+
+//
+// DAO functions
+//
+
+func (impl *DynamicWatchesAnnotationImpl) listAllWatched(namespace string) ([]unstructured.Unstructured, error) {
+	watchedList := &unstructured.UnstructuredList{}
```

Comment:
I know, the big scary frak-your-static-typing `UnstructuredList`. It doesn't matter. We aren't digging through any maps. We do everything through the `metav1.Object` interface, because we really only care about the resource's metadata.